lib/puppet/indirector/file_bucket_file/file.rb

```diff
@@ -38,6 +39,10 @@ def find(request)
     end
 
     def list(request)
+      if request.remote?
+        raise Puppet::Network::AuthorizationError, "Listing remote file buckets is not allowed"
+      end
+
```

Comment:
Puppet has two different ways for doing authz for remote requests. When the HTTP layer receives a request, it will ask the indirection if it allows remote requests:

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/network/http/api/indirected_routes.rb#L39

And the indirection will delegate to its current terminus:

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/indirection.rb#L176-L178

The default terminus behavior is to allow remote access: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/terminus.rb#L144

But in some cases we don't, e.g. keys should never be accessed remotely: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/key/file.rb#L10-L12

However, this mechanism doesn't pass the request in, so the terminus can't decide based on request options.

The other mechanism is for the terminus to call `request.remote?` as is done in this PR. The compiler terminus does something similar for `use_node`: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/catalog/compiler.rb#L131-L133

That said the former approach seems preferable since it's guaranteed that the request came in through the HTTP layer. The latter approach is more indirect as it assumes the request is remote if it has a `node` or `ip`.

Ideally `IndirectedRoutes` would call `Indirection#allow_remote_requests?` passing in the request. However, I don't see how to do that in a backwards compatible way, since people may have written termini that expect no parameters.

One option might be to create new methods on indirection and terminus that take a request parameter, `allow_remote?(request)`, deprecate the old methods, and switch them over to call the new methods. That seems more than I want to do now.

So I think we need to go with the approach in this PR.